src/transport/PeerMessageCounter.h

```diff
@@ -135,6 +191,64 @@ class PeerMessageCounter
         }
     }
 
+    /**
+     * @brief
+     *    With the counter verified and the packet MIC also verified by the secure key, we can trust the packet and adjust
+     *    counter states including possible Rollover for Groups communications.
+     *
+     * @pre Verify(counter) == CHIP_NO_ERROR
+     */
+    void CommitWithRollOver(uint32_t counter)
+    {
+        if (counter <= mSynced.mMaxCounter)
+        {
+            // Take into account rollover
+            uint32_t offset         = mSynced.mMaxCounter - counter;
+            uint32_t rollOverOffset = ((UINT32_MAX - mSynced.mMaxCounter) + counter);
+            if (offset < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+            {
+                mSynced.mWindow.set(offset);
+            }
+            else if (rollOverOffset < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+            {
+                // Roll over occured
+                mSynced.mMaxCounter = counter;
+                mSynced.mWindow <<= rollOverOffset;
+                mSynced.mWindow.set(0);
+            }
+            else
+            {
+                // Roll over occured with a leap bigger than CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE
+                mSynced.mMaxCounter = counter;
+                mSynced.mWindow.reset();
+                mSynced.mWindow.set(0);
+            }
+        }
+        else
+        {
+            uint32_t offset = mSynced.mMaxCounter - counter;
+            if (offset < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+            {
+                mSynced.mWindow.set(offset);
+            }
+            else
+            {
+                // Not a bit inside the window.  Since we are committing, this is a new mMaxCounter value.
+                mSynced.mMaxCounter = counter;
+                uint32_t shift      = -offset;
+                if (shift > CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+                {
+                    mSynced.mWindow.reset();
+                }
+                else
+                {
+                    mSynced.mWindow <<= shift;
+                }
+                mSynced.mWindow.set(0);
```

Comment:
Again, this is not consistent with what `VerifyGroupCounter` does.  We need to either change that code to just do `if (mSynced.mWindow.test(offset))` or change this code to be consistent with it.  And clearly need a test.

Reply:
Code is now consistent with `test(offset)`